csrc/multidevice/utils.cpp

```diff
@@ -192,4 +192,19 @@ std::set<DeviceIdxType> involvedDevices(Expr* expr) {
   return ret;
 }
 
+std::vector<int64_t> unshardedSize(
+    TensorView* tv,
+    c10::IntArrayRef sharded_sizes) {
+  std::vector<int64_t> unsharded_sizes;
+  std::copy(
+      sharded_sizes.begin(),
+      sharded_sizes.end(),
+      std::back_inserter(unsharded_sizes));
+  if (isSharded(tv)) {
+    int num_devices = tv->getDeviceMesh().vector().size();
+    unsharded_sizes[0] = num_devices;
```

Comment:
Please leave a note that:

- The outermost dimension is assumed to be the only sharded domain if the tensor is sharded
- The extent of the sharded dimension is assumed to be equal to the number of devices of the mesh

Both are not necessarily true in general.